apps/designer/app/shared/nano-states/nano-states.ts

```diff
@@ -42,6 +44,40 @@ export const instancesIndexStore = computed(
   }
 );
 
+export const propsStore = atom<Props>([]);
+export const propsIndexStore = computed(propsStore, (props) => {
+  const propsByInstanceId = new Map<Instance["id"], Props>();
+  for (const prop of props) {
+    const { instanceId } = prop;
+    let instanceProps = propsByInstanceId.get(instanceId);
+    if (instanceProps === undefined) {
+      instanceProps = [];
+      propsByInstanceId.set(instanceId, instanceProps);
+    }
+    instanceProps.push(prop);
+  }
+  return {
+    propsByInstanceId,
+  };
+});
+export const useSetProps = (props: Props) => {
+  useSyncInitializeOnce(() => {
+    propsStore.set(props);
+  });
+};
+export const useInstanceProps = (instanceId: undefined | Instance["id"]) => {
+  const instancePropsStore = useMemo(() => {
+    return shallowComputed([propsIndexStore], (propsIndex) => {
+      if (instanceId === undefined) {
+        return [];
+      }
+      return propsIndex.propsByInstanceId.get(instanceId) ?? [];
+    });
+  }, [instanceId]);
+  const instanceProps = useStore(instancePropsStore);
+  return instanceProps;
```

Comment:
Does `propsIndexStore` get recomputed only when `propsStore` change? 

Looking at this code it seems like we might be creating separate instances of the computed store per `instanceId`.

Did you test how this works in practice?

Reply:
Yes, propsIndexStore is recomputed every time propsStore is changed. Each instancePropsStore is recomputed when propsIndexStore is changed. Though useStore triggers rerender only when one of props array elements are changed.

Ideally we need to maintain a separate state with index though it's complicated with patches so we recompute shared index for all invocations instead.